diagnostic/engine.go

```diff
@@ -144,112 +236,6 @@ func (c *conflict) addSimilarConflict(conflict conflict) {
-func (l *ConflictList) AddOverconstraintConflict(nilExplanation ExplainedBool, nonnilExplanation ExplainedBool, pass *analysis.Pass) {
-	c := conflict{}
-
-	// Build nil path by traversing the inference graph from `nilExplanation` part of the overconstraint failure.
-	// (Note that this traversal gives us a backward path from point of conflict to the source of nilability. Hence, we
-	// must take this into consideration while printing the flow, which is currently being handled in `addNilPathNode()`.)
-	var queue []ExplainedBool
-	queue = append(queue, nilExplanation)
-
-	for len(queue) > 0 {
-		e := queue[0]
-		queue = queue[1:]
-
-		t := e.getPrimitiveFullTrigger()
-		// We have two cases here:
-		// 1. No annotation present (i.e., full inference): we have producer and consumer explanations available; use them directly
-		// 2: Annotation present (i.e., no inference): we construct the explanation from the annotation string
-		if t.ConsumerRepr != nil && t.ProducerRepr != nil {
-			c.flow.addNilPathNode(t.ProducerRepr, t.ConsumerRepr)
-		} else {
-			c.flow.addNilPathNode(annotation.LocatedPrestring{
-				Contained: e,
-				Location:  util.TruncatePosition(pass.Fset.Position(t.Pos)),
-			}, nil)
-		}
-
-		if b := e.deeperReason(); b != nil {
-			queue = append(queue, b)
-		}
-	}
-
-	// Build nonnil path by traversing the inference graph from `nonnilExplanation` part of the overconstraint failure.
-	// (Note that this traversal is forward from the point of conflict to dereference. Hence, we don't need to make
-	// any special considerations while printing the flow.)
-	// Different from building the nil path above, here we also want to deduce the position where the error should be reported,
-	// i.e., the point of dereference where the nil panic would occur. In NilAway's context this is the last node
-	// in the non-nil path. Therefore, we keep updating `c.pos` until we reach the end of the non-nil path.
-	queue = make([]ExplainedBool, 0)
-	queue = append(queue, nonnilExplanation)
-	for len(queue) > 0 {
-		e := queue[0]
-		queue = queue[1:]
-
-		t := e.getPrimitiveFullTrigger()
-		// Similar to above, we have two cases here:
-		// 1. No annotation present (i.e., full inference): we have producer and consumer explanations available; use them directly
-		// 2: Annotation present (i.e., no inference): we construct the explanation from the annotation string
-		if t.ConsumerRepr != nil && t.ProducerRepr != nil {
-			c.flow.addNonNilPathNode(t.ProducerRepr, t.ConsumerRepr)
-			c.pos = t.Pos
-		} else {
-			c.flow.addNonNilPathNode(annotation.LocatedPrestring{
-				Contained: e,
-				Location:  util.TruncatePosition(pass.Fset.Position(t.Pos)),
-			}, nil)
-			c.pos = t.Pos
-		}
-
-		if b := e.deeperReason(); b != nil {
-			queue = append(queue, b)
-		}
-	}
-
-	l.conflicts = append(l.conflicts, c)
-}
-
-// Diagnostics returns a list of diagnostics for the conflicts in the list.
-func (l *ConflictList) Diagnostics() []analysis.Diagnostic {
```

Comment:
This now takes a `grouping bool` parameter to indicate whether the generated diagnostics should be grouped or not.